src/zope/interface/_zope_interface_coptimizations.c

```diff
@@ -275,6 +276,32 @@ inst_attr(PyObject *self, PyObject *name)
   PyErr_SetObject(PyExc_AttributeError, name);
   return NULL;
 }
+#else
+/*
+  PyPy.
+  _PyObject_GetDictPtr is a private CPython API and
+  using it on PyPy doesn't work as expected. We must use the documented
+  APIs. This has some subtle differences, notably it would use descriptors.
+  But the tests pass.
+*/
+static PyObject*
+inst_attr(PyObject* self, PyObject* name)
+{
+  PyObject* result;
+  result = PyObject_GetAttr(self, name);
+  if (result != NULL) {
+    /*
+        The CPython version returns a borrowed reference.
+        We don't have that ability with the standard API,
+        so we decref here to mimic it. That should be fine if the
+        attribute was really in the dictionary, but it could be an
+        issue if it was a new object from a descriptor.
```

Comment:
I agree, it scares me too. I think it can be cleaned up nicely, but the tl;dr is that I'd like to deal with that in a follow-up PR. 

There are only four uses of `inst_attr`. I *think* it should be possible to eliminate all of them in favor of inlined fields in the structures, instead of putting them in the `__dict__` directly, if we use `tp_members` or `tp_getset`. That should preserve the suggested speed advantages of `inst_attr`, while also being mostly compatible: 

- Attributes could be read/write accessible to C and Python;
- The attributes would still be listed in `dir`;
- These are private types so we don't have to worry about extensions at the C level;
- If anyone is accessing these attributes from C, they should be using `PyObject_GetAttr` (if they're trying to pull the same trick with the undocumented private API `_PyObject_GetDictPtr`, they'll lose…but eh, undocumented private API; I didn't find any uses of that in other zopefoundation projects, notably not zope.security or zope.proxy)

The major difference is they won't show up in `vars()` or `__dict__`. We'd move them to `__slots__` in the Python implementation to mimic that. We may be able to stop most instances of `Specification(Base)` and especially `ClassProvides(Base)` from even having dicts; given how common those are that could be a nice memory win.

If that doesn't work out, I would propose adding a companion macro, something like `inst_attr_decref` that's required to be called. On PyPy it would expand to a real `DECREF`, on CPython a no-op (I'd also want to add a branch for Python 3 to use the supported API to get an object's dict).

Thoughts?

Reply:
I'm a little less scared now.

I'm not quite sure whether it is possible to trigger a refcounting problem on pypy by writing pure python code that invokes zope.interface's C optimizations somehow?  If this is an internal function that works on instances of internal classes, then everything should be fine, no?

Reply:
I actually think it would be quite hard to trigger some sort of refcounting-type issue here. IIUC, it should be pretty much impossible. 

By definition, the object we're accessing is a strongly held attribute of another strongly held object (i.e., in CPython it's in the object's `__dict__`; it PyPy it's possibly not in a literal `dict`, but part of the hidden class backing the object implementation). According to [this PyPy blog post](https://morepypy.blogspot.com/2018/09/inside-cpyext-why-emulating-cpython-c.html), as long as a particular object exists in the Python/PyPy world, the `PyObject*` that points to it will never go invalid, *even if the refcount is 0*:

> We now need a way to convert between W_Root and PyObject* and vice-versa; also, we need to to ensure that the lifetime of the two entities are in sync. In particular:
> 1. as long as the W_Root is kept alive by the GC, we want the PyObject* to live ((even if its refcount drops to 0**;
> 2. as long as the PyObject* has a refcount greater than 0, we want to make sure that the GC does not collect the W_Root.